File: api/ledger/ledger-utxo/src/main/java/net/corda/v5/ledger/utxo/StateAndRef.java
package net.corda.v5.ledger.utxo;

import net.corda.v5.base.annotations.CordaSerializable;
import net.corda.v5.base.annotations.DoNotImplement;
import org.jetbrains.annotations.NotNull;

/**
 * Defines a composition of a {@link TransactionState} and a {@link StateRef}.
 */
@DoNotImplement
@CordaSerializable
public interface StateAndRef<T extends ContractState> {

    /**
     * Gets the {@link TransactionState} component of the current {@link StateAndRef}.
     *
     * @return Returns the {@link TransactionState} component of the current {@link StateAndRef}.
     */
    @NotNull
    TransactionState<T> getState();

    /**
     * Gets the {@link StateRef} component of the current {@link StateAndRef}.
     *
     * @return Returns the {@link StateRef} component of the current {@link StateAndRef}.
     */
    @NotNull
    StateRef getRef();
}

File: ledger/ledger-utxo/src/main/kotlin/net/corda/v5/ledger/utxo/StateAndRef.kt
package net.corda.v5.ledger.utxo

import net.corda.v5.base.annotations.CordaSerializable
import net.corda.v5.base.annotations.DoNotImplement

/**
 * Represents a composition of a [TransactionState] and a [StateRef].
 *
 * @property state The [TransactionState] component of the current [StateAndRef].
 * @property ref The [StateRef] component of the current [StateAndRef].
 */
@DoNotImplement
@CordaSerializable
interface StateAndRef<out T : ContractState> {
    val state: TransactionState<T>
    val ref: StateRef
}
